packages/opentelemetry-plugin-mysql/src/mysql.ts

```diff
@@ -0,0 +1,266 @@
+      );
+      return function getConnection(
+        arg1?: unknown,
+        arg2?: unknown,
+        arg3?: unknown
+      ) {
+        if (arguments.length === 1 && typeof arg1 === 'function') {
+          const patchFn = thisPlugin._getConnectionCallbackPatchFn(arg1);
+          return originalGetConnection.call(pool, patchFn);
+        }
+        if (arguments.length === 2 && typeof arg2 === 'function') {
+          const patchFn = thisPlugin._getConnectionCallbackPatchFn(arg2);
+          return originalGetConnection.call(pool, arg1, patchFn);
+        }
+        if (arguments.length === 3 && typeof arg3 === 'function') {
+          const patchFn = thisPlugin._getConnectionCallbackPatchFn(arg3);
+          return originalGetConnection.call(pool, arg1, arg2, patchFn);
+        }
+
+        return originalGetConnection.apply(pool, arguments);
+      };
+    };
+  }
+
+  private _getConnectionCallbackPatchFn(cb: Function) {
+    const thisPlugin = this;
+    return function() {
+      if (arguments[1]) {
+        shimmer.wrap(
+          arguments[1],
+          'query',
+          thisPlugin._patchQuery(arguments[1])
+        );
+      }
+      if (typeof cb === 'function') {
+        cb(...arguments);
+      }
+    };
+  }
+
+  private _patchQuery(connection: mysqlTypes.Connection | mysqlTypes.Pool) {
+    return (originalQuery: Function): mysqlTypes.QueryFunction => {
+      const thisPlugin = this;
+      thisPlugin._logger.debug('MysqlPlugin: patched mysql query');
+
+      return function query(
+        query: string | mysqlTypes.Query | mysqlTypes.QueryOptions,
+        _valuesOrCallback?: unknown[] | mysqlTypes.queryCallback,
+        _callback?: mysqlTypes.queryCallback
+      ) {
+        const spanName = getSpanName(query);
+
+        const span = thisPlugin._tracer.startSpan(spanName, {
+          parent: thisPlugin._tracer.getCurrentSpan() || undefined,
+          kind: SpanKind.CLIENT,
+          attributes: {
+            ...MysqlPlugin.COMMON_ATTRIBUTES,
+            ...getConnectionAttributes(connection.config),
+          },
+        });
+
+        if (typeof query === 'string') {
+          span.setAttribute(AttributeNames.DB_STATEMENT, query);
+        } else if (typeof query === 'object') {
+          if (query.sql) {
+            span.setAttribute(AttributeNames.DB_STATEMENT, query.sql);
+          }
+
+          if (query.values) {
+            span.setAttribute(AttributeNames.MYSQL_VALUES, query.values);
+          }
+        }
+
+        if (arguments.length === 1) {
+          const streamableQuery: mysqlTypes.Query = originalQuery.apply(
+            connection,
+            arguments
+          );
+
+          return streamableQuery
+            .on('error', err =>
+              span.setStatus({
+                code: CanonicalCode.UNKNOWN,
+                message: err.message,
+              })
+            )
+            .on('end', () => {
+              span.end();
```

Comment:
If the `error` event happens, will the `end` event happen as well? Also will this span have a status by default? (Is `OK` set somewhere else?)

Reply:
Is OK not the default? In examples/mysql the OK status code is properly set in spite of never being explicitly set.